stdlib/public/core/UnsafeRawPointer.swift.gyb

```diff
@@ -0,0 +1,557 @@
+  ///
+  /// - Postcondition: The `T` values at `self..<self + source.count *
+  ///   strideof(T.self)` are initialized.
+  ///
+  /// TODO: Optimize where `C` is a `ContiguousArrayBuffer`.
+  @discardableResult
+  public func initializeMemory<C : Collection>(
+    as: C.Iterator.Element.Type, from source: C
+  ) -> UnsafeMutablePointer<C.Iterator.Element> {
+    // Initialize and bind each element of the container.
+    for (index, element) in source.enumerated() {
+      self.initializeMemory(as: C.Iterator.Element.self, at: index, to: element)
+    }
+    return UnsafeMutablePointer(_rawValue)
+  }
+
+  /// Initialize memory starting at `self` with `count` `T` values
+  /// beginning at `source`, bind the initialized memory to type `T`,
+  /// and return the source memory to an uninitialized state.
+  ///
+  /// Returns an `UnsafeMutablePointer<T>` this memory.
+  ///
+  /// - Precondition: `count >= 0`
+  ///
+  /// - Precondition: The memory at `self..<self + count *
+  ///   strideof(T.self)` is uninitialized and the `T` values at
+  ///   `source..<source + count` are initialized.
+  ///
+  /// - Postcondition: The memory at `self..<self + count *
+  ///   strideof(T.self)` is bound to type `T`.
+  ///
+  /// - Postcondition: The `T` values at `self..<self + count *
+  ///   strideof(T.self)` are initialized and the memory at
+  ///   `source..<source + count` is uninitialized.
+  @discardableResult
+  public func moveInitializeMemory<T>(
+    as type: T.Type, from source: UnsafeMutablePointer<T>, count: Int
+  ) -> UnsafeMutablePointer<T> {
+    _debugPrecondition(
+      count >= 0,
+      "UnsafeMutableRawPointer.moveInitializeMemory with negative count")
+
+    Builtin.bindMemory(_rawValue, count._builtinWordValue, type)
+    if self < UnsafeMutableRawPointer(source)
+       || self >= UnsafeMutableRawPointer(source + count) {
+      // initialize forward from a disjoint or following overlapping range.
+      Builtin.takeArrayFrontToBack(
+        T.self, self._rawValue, source._rawValue, count._builtinWordValue)
+      // This builtin is equivalent to:
+      // for i in 0..<count {
+      //   (self.assumingMemoryBound(to: T.self) + i)
+      //   .initialize(to: (source + i).move())
+      // }
+    }
+    else {
+      // initialize backward from a non-following overlapping range.
+      Builtin.takeArrayBackToFront(
+        T.self, self._rawValue, source._rawValue, count._builtinWordValue)
+      // This builtin is equivalent to:
+      // var src = source + count
+      // var dst = self.assumingMemoryBound(to: T.self) + count
+      // while dst != self {
+      //   (--dst).initialize(to: (--src).move())
+      // }
+    }
+    return UnsafeMutablePointer(_rawValue)
+  }
+%  end # mutable
+
+  /// Read raw bytes from memory at `self + offset` and construct a
+  /// value of type `T`.
+  ///
+  /// - Precondition: The underlying pointer plus `offset` is properly
+  ///   aligned for accessing `T`.
```

Comment:
Did you consider adding a debug precondition for alignment?  I expect this to be frequently violated.